### ISSUE TITLE
amp-carousel/amp-base-carousel: No indication of focus

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#31171]
## WCAG Level
Level AA

## Priority
High

## Type
Navigation

## Pages/screens/components affected
* [amp-base-carousel](https://amp.dev/documentation/components/amp-base-carousel/)
* amp-stream-gallery
* [amp-inline-gallery](https://amp.dev/documentation/components/amp-inline-gallery/)
* [amp-lightbox-gallery](https://amp.dev/documentation/components/amp-lightbox-gallery/)

## Description
Keyboard users rely on the browser's focus outline to keep track of where they are on web pages. The currently selected form input, button, or hyperlink is the one said to have the focus.

Most browsers use a faint dotted outline to indicate the focus to the user, though some browsers like Chrome or Safari use an enhanced focus outline, consisting of a coloured border around the selected link. It is also possible for developers to provide custom focus indications within the site's cascading style sheets (CSS).

The standard previous/next controls for `amp-base-carousel` and the image gallery variants (which are all based on the carousel) currently lack a visible indication of focus. The browser's default outline has been suppressed, and no alternative has been provided.

```
.i-amphtml-base-carousel-arrow {
  ...
  outline: none;
  ...
}
```

In addition, the controls are visually hidden after a user has interacted with the carousel/gallery using touch, or if an author has set the `controls="never"` attribute (note: this feature is not currently documented, but visible in the code for the [`amp-base-carousel` hidden controls demo](https://nutritious-tide-hornet.glitch.me/amp-base-carousel/hidden-controls.html)). The buttons are still focusable, and can still be reached by keyboard users, but they remain completely invisible - while also still not providing any visible indication of focus.

![Screenshot of an image gallery, when first loaded - showing visible previous/next buttons - and after a touch interaction - with the controls now hidden, but still present in markup and focusable/operable with the keyboard](https://user-images.githubusercontent.com/54933393/99181801-dcb5a900-2728-11eb-8166-20820b85ac54.png)

Once visually hidden, these controls are then also completely unusable for mouse users (as they're visually hidden, and their mouse interactions are suppressed via CSS).

```
amp-base-carousel[i-amphtml-base-carousel-hide-buttons] .i-amphtml-base-carousel-arrow-next-slot>*,
amp-base-carousel[i-amphtml-base-carousel-hide-buttons] .i-amphtml-base-carousel-arrow-prev-slot>*,
amp-base-carousel[loop=false] .i-amphtml-base-carousel-arrow-next-slot>[disabled],
amp-base-carousel[loop=false] .i-amphtml-base-carousel-arrow-prev-slot>[disabled] {
  opacity: 0;
  pointer-events: none;
}
```

Note that this aspect (inability to use mouse after touch interaction) falls outside of WCAG 2.1 Level A/AA (which this evaluation was based on), but does fail [WCAG 2.1 Success Criterion 2.5.6 Concurrent Input Mechanisms (Level AAA)](https://www.w3.org/TR/WCAG/#concurrent-input-mechanisms).

## User impact
Users who navigate using the keyboard will have difficulty operating the carousel/galleries, as it will not be clear to these users when the previous/next controls have focus. This problem will be further exacerbated in situations where the controls have been hidden (either due to an interaction using touch, or because the author has explicitly hidden them), as keyboard users will essentially see their focus "disappear" completely when they reach the invisible controls.

The default behaviour of permanently hiding the controls when the carousel/gallery has been operated via touch does not take into account users who may be using a touchscreen in conjunction with other input methods, such as a keyboard or mouse. They may be on a laptop/desktop with a touchscreen (e.g. Surface laptop), or use a "touch" device such as a smartphone or tablet, but have a paired Bluetooth mouse and/or keyboard. For users who occasionally use the touchscreen, but then switch to any of their other input mechanisms, the carousel/gallery will become unusable unless they, after the first touchscreen interaction, consistently only use touch to operate the carousels.

## Required solution
By default, the previous/next buttons must provide a visible focus indication. We recommend providing an explicit `:focus` style for these controls that is clear and visible - for instance, a stronger/thicker stroke/border.

Note that the `:focus` styles will also take effect after a mouse user activated one of the buttons (which is likely why no focus styles have been implemented in the first place). To avoid this (as it may be visually confusing to mouse users), consider using the new [`:focus-visible` pseudo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible) in combination with the [`:focus-visible` polyfill](https://github.com/WICG/focus-visible).

While it may be appropriate to visually hide the previous/next buttons when the user interacts with the page using a touchscreen, this should be reverted - and the buttons be made visible - as soon as the user switches to using a mouse or keyboard again. This same logic should also be applied when the author explicitly decided to hide the controls with `controls="never"`.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Navigate through the page using a keyboard (`Tab`/`Shift`+`Tab`)
2. Verify that controls that receive focus have a visible focus indication.
3. Verify that the keyboard focus does not "disappear" to any invisible controls
4. Interact with the page/carousel/gallery using a touchscreen. Wait for the previous/next buttons to disappear
5. Switch to the keyboard or mouse
6. Verify that the previous/next controls become visible again, and that focus does not "disappear" (i.e. that the focus doesn't end up on invisible elements again)

## Related standards
* [WCAG 2.1 Success Criterion 2.4.7 Focus Visible (Level AA)](https://www.w3.org/TR/WCAG21/#focus-visible)
* [WCAG 2.1 Success Criterion 2.5.6 Concurrent Input Mechanisms (Level AAA)](https://www.w3.org/TR/WCAG21/#concurrent-input-mechanisms)

## More information
* [W3C - Failure of Success Criterion 2.4.7 due to styling element outlines and borders in a way that removes or renders non-visible the visual focus indicator](https://www.w3.org/WAI/WCAG21/Techniques/failures/F78.html)

## Test data
Test date: November 2020
